src/core/algorithms/fd/fd_verifier/dynamic_fd_verifier.cpp

```diff
@@ -0,0 +1,243 @@
+        throw std::runtime_error("Got an empty dataset: FD verifying is meaningless.");
+    }
+    CreateFD();
+    input_table_->Reset();
+    stats_calculator_ =
+            std::make_unique<DynamicStatsCalculator>(input_table_, lhs_indices_, rhs_indices_);
+    VerifyFD();
+    SortHighlightsByProportionDescending();
+}
+
+unsigned long long DynamicFDVerifier::ExecuteInternal() {
+    auto start_time = std::chrono::system_clock::now();
+    std::vector<std::pair<std::optional<size_t>, std::vector<int>>> lhs_inserts{}, rhs_inserts{};
+    std::unordered_set<size_t> deletes_and_updates_indices{delete_statenents_indicies_};
+    if (insert_statements_table_ != nullptr) {
+        while (insert_statements_table_->HasNextRow()) {
+            std::vector<std::string> row = insert_statements_table_->GetNextRow();
+            if (row.size() != input_table_->GetNumberOfColumns()) {
+                LOG(WARNING) << "Received row with size " << row.size() << ", but expected "
+                             << input_table_->GetNumberOfColumns();
+                continue;
+            }
+            lhs_inserts.emplace_back(std::nullopt, ParseRowForPLI(row.begin(), lhs_indices_));
+            rhs_inserts.emplace_back(std::nullopt, ParseRowForPLI(row.begin(), rhs_indices_));
+        }
+        insert_statements_table_->Reset();
+    }
+    if (update_statements_table_ != nullptr) {
+        while (update_statements_table_->HasNextRow()) {
+            std::vector<std::string> row = update_statements_table_->GetNextRow();
+            if (row.size() != input_table_->GetNumberOfColumns() + 1) {
+                LOG(WARNING) << "Received row with size " << row.size() << ", but expected "
+                             << input_table_->GetNumberOfColumns() + 1;
+                continue;
+            }
+            size_t row_id = std::stoull(row[0]);
+            lhs_inserts.emplace_back(row_id, ParseRowForPLI(row.begin() + 1, lhs_indices_));
+            rhs_inserts.emplace_back(row_id, ParseRowForPLI(row.begin() + 1, rhs_indices_));
+            deletes_and_updates_indices.emplace(row_id);
+        }
+        update_statements_table_->Reset();
+    }
+
+    lhs_pli_->UpdateWith(lhs_inserts, deletes_and_updates_indices);
+    rhs_pli_->UpdateWith(rhs_inserts, deletes_and_updates_indices);
+    stats_calculator_->UpdateTableData(insert_statements_table_, update_statements_table_,
+                                       delete_statenents_indicies_);
+
+    VerifyFD();
+    SortHighlightsByProportionDescending();
+
+    auto elapsed_milliseconds = std::chrono::duration_cast<std::chrono::milliseconds>(
+            std::chrono::system_clock::now() - start_time);
+    return elapsed_milliseconds.count();
+}
+
+void DynamicFDVerifier::VerifyFD() const {
+    std::unique_ptr<model::DynPLI const> intersection_pli = lhs_pli_->Intersect(rhs_pli_.get());
+
+    if (lhs_pli_->GetNumCluster() == intersection_pli->GetNumCluster()) {
+        return;
```

Comment:
Statistics might have irrelevant values if they aren't updated. It makes sense for `get_error` to return `0.0` if the FD holds, for example.

Reply:
this case handles during ResetState() call

Reply:
Please add a comment for that, then.
Isn't there going to be an optimized version that won't have to redo all the calculations, by the way? I'd expect it to be updating statistics directly. You should probably move resetting statistics here instead, since it is kind of a crutch before the "proper" method is implemented. That is, statistics are going to be reused, not reset every time.